core/templates/components/graph-services/graph-layout.service.spec.ts

```diff
@@ -0,0 +1,1251 @@
+      State5: {
+        depth: 1,
+        offset: 3.5,
+        reachable: true,
+        x0: 0.8600000000000001,
+        y0: 0.26,
+        xLabel: 0.9500000000000001,
+        yLabel: 0.30000000000000004,
+        id: 'State5',
+        label: 'State5',
+        height: 0.08,
+        width: 0.18000000000000002,
+        reachableFromEnd: false
+      },
+      State6: {
+        depth: 2,
+        offset: 1.5,
+        reachable: true,
+        x0: 0.41000000000000003,
+        y0: 0.3933333333333333,
+        xLabel: 0.5,
+        yLabel: 0.43333333333333335,
+        id: 'State6',
+        label: 'State6',
+        height: 0.08,
+        width: 0.18000000000000002,
+        reachableFromEnd: false
+      },
+      State7: {
+        depth: 2,
+        offset: 2.5,
+        reachable: true,
+        x0: 0.6350000000000001,
+        y0: 0.3933333333333333,
+        xLabel: 0.7250000000000001,
+        yLabel: 0.43333333333333335,
+        id: 'State7',
+        label: 'State7',
+        height: 0.08,
+        width: 0.18000000000000002,
+        reachableFromEnd: false
+      },
+      State8: {
+        depth: 3,
+        offset: 0,
+        reachable: true,
+        x0: 0.07250000000000001,
+        y0: 0.5266666666666667,
+        xLabel: 0.1625,
+        yLabel: 0.5666666666666667,
+        id: 'State8',
+        label: 'State8',
+        height: 0.08,
+        width: 0.18000000000000002,
+        reachableFromEnd: true
+      },
+      State9: {
+        depth: 1,
+        offset: 0.5,
+        reachable: true,
+        x0: 0.185,
+        y0: 0.26,
+        xLabel: 0.275,
+        yLabel: 0.30000000000000004,
+        id: 'State9',
+        label: 'State9',
+        height: 0.08,
+        width: 0.18000000000000002,
+        reachableFromEnd: true
+      },
+      Orphaned: {
+        depth: 5,
+        offset: 0,
+        reachable: false,
+        x0: 0.07250000000000001,
+        y0: 0.7933333333333333,
+        xLabel: 0.1625,
+        yLabel: 0.8333333333333333,
+        id: 'Orphaned',
+        label: 'Orphaned',
+        height: 0.08,
+        width: 0.18000000000000002,
+        reachableFromEnd: false
+      }
+    };
+    let nodeData1 = cloneDeep(nodeData);
+
+    let graphWidthUpperBound = sgls.getGraphWidth(
+      AppConstants.MAX_NODES_PER_ROW, AppConstants.MAX_NODE_LABEL_LENGTH);
+    let graphHeight = sgls.getGraphHeight(nodeData1);
+
+    // Here, modifiedNodeData is nodeData with position values in pixels.
+    // For ex, nodeData.State1.x0 = 0.07250000000000001 which is expressed
+    // in fraction of graph width. It can be converted to pixels by multiplying
+    // with graph width. So, modifiedNodeDate.State1.x0 = 0.07250000000000001 *
+    // graphWidthUpperBound (630) = 47.675.
+    let modifiedNodeData = sgls.modifyPositionValues(
+      nodeData, graphWidthUpperBound, graphHeight);
+
+    // Verifying the position values of State1.
+    expect(modifiedNodeData.State1.x0).toEqual(nodeData.State1.x0);
```

Comment:
The way this is written is a bit confusing because modifiedNodeData is apparently the same as nodeData. So modifyPositionValues() looks like it is basically equivalent to a function that doesn't do anything at all. This test is therefore rather weak. Can you make it stronger so that it can actually distinguish an incorrect implementation of modifyPositionValues()?

Reply:
Ah, I just realised I was checking modifiedNodeData against nodeData (which also is modified). 

Changed the test